pkg/agent/trustbundlesources/bundle.go

```diff
@@ -8,52 +8,230 @@ import (
 )
 
-func SetupTrustBundle(ac *agent.Config, bconfig *Config) error {
-	// Either download the trust bundle if TrustBundleURL is set, or read it
-	// from disk if TrustBundlePath is set
-	ac.InsecureBootstrap = bconfig.InsecureBootstrap
+type Bundle struct {
+	config             *Config
+	use                int
+	connectionAttempts int
+	startTime          time.Time
+	log                logrus.FieldLogger
+	metrics            telemetry.Metrics
+	storage            storage.Storage
+	lastBundle         []*x509.Certificate
+}
+
+func New(config *Config, log logrus.FieldLogger) *Bundle {
+	return &Bundle{
+		config: config,
+		log:    log,
+	}
+}
+
+func (b *Bundle) SetMetrics(metrics telemetry.Metrics) {
+	b.metrics = metrics
+}
+
+func (b *Bundle) SetStorage(sto storage.Storage) error {
+	b.storage = sto
+	use, startTime, connectionAttempts, err := b.storage.LoadBootstrapState()
+	b.use = use
+	b.startTime = startTime
+	b.connectionAttempts = connectionAttempts
+	if use == UseUnspecified {
+		b.use = UseBootstrap
+		BootstrapTrustBundle, err := b.storage.LoadBundle()
+		if err != nil {
+			if !errors.Is(err, storage.ErrNotCached) {
+				return err
+			}
+			b.use = UseBootstrap
+		} else if len(BootstrapTrustBundle) > 0 {
+			b.use = UseRebootstrap
+		}
+	}
+	b.updateMetrics()
+	return err
+}
+
+func (b *Bundle) SetUse(use int) error {
+	if b.use != use {
+		b.use = use
+		b.connectionAttempts = 0
+		b.startTime = time.Now()
+		b.log.Info("Setting use.")
+		err := b.storage.StoreBootstrapState(use, b.startTime, b.connectionAttempts)
+		if err != nil {
+			return err
+		}
+		b.updateMetrics()
+		return err
+	}
+	return nil
+}
+
+func (b *Bundle) SetSuccessIfRunning() error {
+	if !b.startTime.IsZero() {
+		return b.SetSuccess()
+	}
+	return nil
+}
+
+func (b *Bundle) SetSuccess() error {
+	var err error
+	b.log.Info(fmt.Sprintf("Success after %s attempts=%d", time.Since(b.startTime), b.connectionAttempts))
+	b.use = UseRebootstrap
+	b.connectionAttempts = 0
+	b.startTime = time.Time{}
+	b.log.Info("Setting use.")
+	if b.storage != nil {
+		err = b.storage.StoreBootstrapState(b.use, b.startTime, b.connectionAttempts)
+		if err != nil {
+			return err
+		}
+		b.updateMetrics()
+		err = b.storage.StoreBundle(b.lastBundle)
+	}
+	return err
+}
 
+func (b *Bundle) SetForceRebootstrap() error {
+	b.use = UseRebootstrap
+	b.startTime = time.Now()
+	b.connectionAttempts = 0
+	err := b.storage.StoreBootstrapState(b.use, b.startTime, b.connectionAttempts)
+	if err != nil {
+		return err
+	}
+	b.updateMetrics()
+	err = b.storage.DeleteSVID()
+	if err != nil {
+		return err
+	}
+	err = b.storage.StoreBundle(nil)
+	return err
+}
+
+func (b *Bundle) GetStartTime() (time.Time, error) {
+	var err error
+	if b.startTime.IsZero() {
+		b.startTime = time.Now()
+		err = b.storage.StoreBootstrapState(b.use, b.startTime, b.connectionAttempts)
+		b.updateMetrics()
+	}
+	return b.startTime, err
+}
+
+func (b *Bundle) IsBootstrap() bool {
+	return b.use != UseRebootstrap
+}
```

Comment:
None of these functions seem to be covered by tests. Could you add test coverage for them?